### PR TITLE
ffmpeg software decoder implementation

### PR DESCRIPTION
**Pull request by rrawther**
Implemented ffmpeg based software decoder
Modified videoDecode sample to allow ffmpeg based decoding

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@ Full documentation for rocDecode is available at [https://rocm.docs.amd.com/proj
 * Clang is now the default CXX compiler.
 * The new minimum supported version of va-api is 1.16.
 * New build and runtime options have been added to the `rocDecode-setup.py` setup script.
+* Added FFMpeg based software decoding into utils.
+* Modified videodecode sample to allow FFMpeg based decoding
```
Comment (spolifroni-amd) [inline CHANGELOG.md:40]: Do these two lines need to be in the 6.3 release notes?

**Review by spolifroni-amd — COMMENTED** [review]
If the additional lines in the changelogs are for 6.3, please let me know.

File: CHANGELOG.md (hunk 1 above)
Comment (rrawther) [reply CHANGELOG.md:40]: @spolifroni-amd: That will go to 6.4 I think. It doesn't need to be on 6.3 release notes

Comment (spolifroni-amd) [reply CHANGELOG.md:40]: Thanks for clearing that up!

**Review by spolifroni-amd — DISMISSED** [review]

File: utils/ffmpegvideodecode/ffmpeg_video_dec.h [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+        /**
+         *   @brief  This function reconfigure decoder if there is a change in sequence params.
+         */
+        int ReconfigureDecoder(RocdecVideoFormat *p_video_format);
+
+        void DecodeThread();
+        int DecodeAvFrame(AVPacket *av_pkt, AVFrame *p_frame);
+        void InitOutputFrameInfo(AVFrame *p_frame);
+        void PushPacket(AVPacket *pkt){
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.h]: minor formatting: add a space `*pkt){` -> `*pkt) {`

File: utils/ffmpegvideodecode/ffmpeg_video_dec.h [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            {
+                std::lock_guard<std::mutex> lock(mtx_pkt_q_);
+                av_packet_q_.push(pkt);
+            }
+            cv_pkt_.notify_one();
+        }
+        
+        AVPacket *PopPacket(){
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.h]: minor formatting: add a space `*PopPacket(){` -> `*PopPacket() {`

File: utils/ffmpegvideodecode/ffmpeg_video_dec.h [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+            AVPacket *pkt;
+            std::unique_lock<std::mutex> lock(mtx_pkt_q_);
+            cv_pkt_.wait(lock, [&] { return !av_packet_q_.empty(); });
+            pkt = av_packet_q_.front();
+            av_packet_q_.pop();
+            return pkt;
+        }
+
+        void PushFrame(AVFrame *av_frame){
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.h]: minor formatting: add a space `*av_frame){` -> `*av_frame) {`

File: utils/ffmpegvideodecode/ffmpeg_video_dec.h [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+            {
+                std::lock_guard<std::mutex> lock(mtx_frame_q_);
+                av_frame_q_.push(av_frame);
+            }
+            cv_frame_.notify_one();
+        };
+
+        AVFrame *PopFrame(){
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.h]: minor formatting: add a space `*PopFrame(){` -> `*PopFrame() {`

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 6]
```diff
@@ -0,0 +1,866 @@
...
+ * @brief helper function for inferring AVCodecID from rocDecVideoSurfaceFormat
+ * 
+ * @param rocdec_codec 
+ * @return AVCodecID 
+ */
+static inline rocDecVideoSurfaceFormat AVPixelFormat2rocDecVideoSurfaceFormat(AVPixelFormat av_pixel_format) {
+    switch (av_pixel_format) {
+        case AV_PIX_FMT_YUV420P : 
+        case AV_PIX_FMT_YUVJ420P : 
+            return rocDecVideoSurfaceFormat_YUV420;           // need to see if this is actually correct
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: Is this comment still relevant? If so, what needs to be done to ensure we return the correct rocdecode surface format?

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 7]
```diff
@@ -0,0 +1,866 @@
...
+        }
+        // get the output pixel format from dec_context_
+        decoder_pixel_format_ = (dec_context_->pix_fmt == AV_PIX_FMT_NONE) ? AV_PIX_FMT_YUV420P : dec_context_->pix_fmt;
+        dec_context_->thread_count = 4;
+        dec_context_->thread_type = FF_THREAD_FRAME;
+
+    }
+    // allocate av_frame buffer pool for number of surfaces
+    if (dec_frames_.empty()) {
+        for (int i=0; i < num_decode_surfaces; i++) {
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: minor formatting: add spaces `i=0` -> `i = 0`

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 8]
```diff
@@ -0,0 +1,866 @@
...
+            AVFrame *p_frame = av_frame_alloc();
+            dec_frames_.push_back(p_frame);
+        }
+        av_frame_cnt_ = 0;
+    }
+    if (av_packet_data_.empty()) {
+        for (int i=0; i < num_decode_surfaces; i++) {
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: minor formatting: add spaces `i=0` -> `i = 0`

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 9]
```diff
@@ -0,0 +1,866 @@
...
+            uint8_t *pkt_data = static_cast<uint8_t *> (av_malloc(MAX_AV_PACKET_DATA_SIZE));
+            av_packet_data_.push_back(std::make_pair(pkt_data, MAX_AV_PACKET_DATA_SIZE));
+        }
+    }
+
+    // allocate av_packets_ for decoding
+    if (av_packets_.empty()) {
+        for (int i=0; i < num_decode_surfaces; i++) {
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: minor formatting: add spaces `i=0` -> `i = 0`

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 10]
```diff
@@ -0,0 +1,866 @@
...
+
+    if (!(crop_rect_.right && crop_rect_.bottom)) {
+        target_width_ = (disp_width_ + 1) & ~1;
+        target_height_ = (disp_height_ + 1) & ~1;
+    } else {
+        target_width_ = (crop_rect_.right - crop_rect_.left + 1) & ~1;
+        target_height_ = (crop_rect_.bottom - crop_rect_.top + 1) & ~1;
+    }
+
+    chroma_height_ = (int)(ceil(target_height_ * GetChromaHeightFactor(video_surface_format_)));
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: formatting: please use `static_cast<int>(  )` here and in lines 284 and 288 below.

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 11]
```diff
@@ -0,0 +1,866 @@
...
+ */
+int FFMpegVideoDecoder::HandlePictureDecode(RocdecPicParams *pPicParams) {
+    AVPacket *av_pkt = av_packets_[av_pkt_cnt_];
+    std::pair<uint8_t *, int> *packet_data = &av_packet_data_[av_pkt_cnt_];
+    if (last_packet_.payload_size > packet_data->second) {
+        void *new_pkt_data = av_realloc(av_pkt->data, (last_packet_.payload_size + MAX_AV_PACKET_DATA_SIZE));  // add more to avoid frequence reallocation
+        if (!new_pkt_data) {
+            std::cerr << "ERROR: couldn't allocate packet data" << std::endl;
+        }
+        packet_data->first   = static_cast<uint8_t *> (new_pkt_data);
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: minor formatting: remove the space `<uint8_t *> (new_pkt_data);` -> `<uint8_t *>(new_pkt_data);`

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 12]
```diff
@@ -0,0 +1,866 @@
...
+    #else
+        //push packet into packet q for decoding
+        PushPacket(&pkt);
+    #endif
+    }
+
+    if (b_extract_sei_message_) {
+        if (sei_message_display_q_[pDispInfo->picture_index].sei_data) {
+            // Write SEI Message
+            uint8_t *sei_buffer = (uint8_t *)(sei_message_display_q_[pDispInfo->picture_index].sei_data);
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: please use `static_cast<uint8_t *>( )` here

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 13]
```diff
@@ -0,0 +1,866 @@
...
+            free(sei_message_display_q_[pDispInfo->picture_index].sei_data);
+            sei_message_display_q_[pDispInfo->picture_index].sei_data = NULL; // to avoid double free
+            free(sei_message_display_q_[pDispInfo->picture_index].sei_message);
+            sei_message_display_q_[pDispInfo->picture_index].sei_message = NULL; // to avoid double free
+        }
+    }
+    // vp_frames_ffmpeg_.size() is empty, wait for decoding to finish
+    // this will happen during PopFrame()
+    AVFrame *p_av_frame;
+    //std::cout << "Before pop frame " << std::endl;
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: Please remove this comment if it is unnecessary.

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 14]
```diff
@@ -0,0 +1,866 @@
...
+        av_frame_q_.pop();
+#else        
+        p_av_frame = PopFrame();
+#endif        
+
+    if (p_av_frame == nullptr) {
+        std::cerr << "Invalid avframe decode output" << std::endl;
+        return 0;
+    }
+    void * src_ptr[3] = { 0 };
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: minor formatting: remove the space -> `void *` -> `void*`

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 15]
```diff
@@ -0,0 +1,866 @@
...
+    src_pitch[0] = p_av_frame->linesize[0];
+    src_pitch[1] = p_av_frame->linesize[1];
+    src_pitch[2] = p_av_frame->linesize[2];
+
+    // copy the decoded surface info device or host
+    uint8_t *p_dec_frame = nullptr;
+    {
+        std::lock_guard<std::mutex> lock(mtx_vp_frame_);
+        // if not enough frames in stock, allocate
+        if ((unsigned)++output_frame_cnt_ > vp_frames_ffmpeg_.size()) {
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: please use `static_cast` here

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 16]
```diff
@@ -0,0 +1,866 @@
...
+ * 
+ * @param pDispInfo 
+ * @return int 0:fail 1: success
+ */
+int FFMpegVideoDecoder::HandlePictureDisplay(RocdecParserDispInfo *pDispInfo) {
+    // check if we reached eos here. This is a hack since HandlePictureDecode won't be called at the end_of_sequece. 
+    // so we need to flush FFMpeg decoder when we have received the lastpacket with 0 bytes
+    if (!last_packet_.payload_size && !end_of_stream_) {
+        AVPacket pkt = { 0 };
+    #if NO_DECODE_THREAD
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: I recommend using an environment variable to control whether decoding is performed with or without a thread. This change would enable us to test both options without the need to recompile the program. For instance, we could create an environment variable called `ROCDEC_FFMPEG_DISABLE_DECODE_THREAD`. By setting it with the command `export ROCDEC_FFMPEG_DISABLE_DECODE_THREAD=1`, we can disable threading for FFMPEG. The following code can then be used to check whether threading is disabled.

```
const char *is_threading_disabled = std::getenv("ROCDEC_FFMPEG_DISABLE_DECODE_THREAD");

if (std::string(is_threading_disabled) != "1") 
   DecodeAvFrame(&pkt, dec_frames_[av_frame_cnt_]);
else
 PushPacket(&pkt);
```

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 17]
```diff
@@ -0,0 +1,866 @@
...
+            sei_message_display_q_[pDispInfo->picture_index].sei_data = NULL; // to avoid double free
+            free(sei_message_display_q_[pDispInfo->picture_index].sei_message);
+            sei_message_display_q_[pDispInfo->picture_index].sei_message = NULL; // to avoid double free
+        }
+    }
+    // vp_frames_ffmpeg_.size() is empty, wait for decoding to finish
+    // this will happen during PopFrame()
+    AVFrame *p_av_frame;
+    //std::cout << "Before pop frame " << std::endl;
+#if NO_DECODE_THREAD
```
Comment (AryanSalmanpour) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: As mentioned above, it is recommended to use an environment variable to check if threading is disabled.

**Review by AryanSalmanpour — COMMENTED** [review]
I added a few minor formatting and some suggestions.

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp (hunk 6 above)
Comment (rrawther) [reply utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: I think this is correct. I have removed the comment. I thought it was left over when we were using NV12

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp (hunk 17 above)
Comment (jeffqjiangNew) [reply utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: Please make sure all AVC/HEVC/AV1 conformance streams pass with the FFMPEG path before merging. The conformance streams test various corner cases related to resolution change, frame flushing or skipping, frame or field (AVC case), etc., which can be affected by the utility code above FFMPEG AV lib.

Comment (rrawther) [reply utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: @Jeff. This is mainly for cutomer to use FFMpeg decoding option for HEVC. So currently I have only enabled it for that codec. Once we merge to tot develop, we can add all codec support. We will need to add those as future PRs

File: samples/videoDecode/videodecode.cpp [hunk 18]
```diff
@@ -203,18 +214,26 @@ int main(int argc, char **argv) {
     try {
         std::size_t found_file = input_file_path.find_last_of('/');
         std::cout << "info: Input file: " << input_file_path.substr(found_file + 1) << std::endl;
+        RocVideoDecoder *viddec;
```
Comment (LakshmiKumar23) [inline samples/videoDecode/videodecode.cpp:217]: can we initialize to `nullptr`?

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 19]
```diff
@@ -0,0 +1,866 @@
...
+            uint8_t *pkt_data = static_cast<uint8_t *> (av_malloc(MAX_AV_PACKET_DATA_SIZE));
+            av_packet_data_.push_back(std::make_pair(pkt_data, MAX_AV_PACKET_DATA_SIZE));
+        }
+    }
+
+    // allocate av_packets_ for decoding
+    if (av_packets_.empty()) {
+        for (int i=0; i < num_decode_surfaces; i++) {
+            AVPacket *pkt = av_packet_alloc();
+            pkt->data = static_cast<u_int8_t *> (av_packet_data_[i].first);
```
Comment (LakshmiKumar23) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: did you mean `uint8_t`? Not sure what `u_int8_t` is.

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 20]
```diff
@@ -0,0 +1,866 @@
...
+            target_height_ = (disp_height_ + 1) & ~1;
+        } else {
+            target_width_ = (crop_rect_.right - crop_rect_.left + 1) & ~1;
+            target_height_ = (crop_rect_.bottom - crop_rect_.top + 1) & ~1;
+        }
+    }
+
+    surface_stride_ = target_width_ * byte_per_pixel_;
+    chroma_height_ = static_cast<int>(std::ceil(target_height_ * GetChromaHeightFactor(video_surface_format_)));
+    chroma_width_ = (int)(ceil(target_width_ * GetChromaWidthFactor(video_surface_format_)));
```
Comment (LakshmiKumar23) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: please use `static_cast` here

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 21]
```diff
@@ -0,0 +1,866 @@
...
+        output_surface_info_.mem_type = OUT_SURFACE_MEM_HOST_COPIED;
+    }
+}
+
+void FFMpegVideoDecoder::DecodeThread()
+{
+    AVPacket *pkt;
+    do {
+        pkt = PopPacket();
+        //std::cout << "pop packet" << std::endl;
```
Comment (LakshmiKumar23) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: remove comment if unnecessary

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 22]
```diff
@@ -0,0 +1,866 @@
...
+    int status;
+    //send packet to av_codec
+    status = avcodec_send_packet(dec_context_, av_pkt);
+    if (status < 0) {
+        std::cout << "Error sending av packet for decoding: status: " << status << std::endl;
+    }
+    while (status >= 0) {
+        status = avcodec_receive_frame(dec_context_, p_frame);
+        if (status == AVERROR(EAGAIN) || status == AVERROR_EOF) {
+            //av_frame_free(&p_frame);
```
Comment (LakshmiKumar23) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: remove line if unnecessary

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 23]
```diff
@@ -0,0 +1,866 @@
...
+        }
+        else if (status < 0) {
+            std::cout << "Error during decoding" << std::endl;
+            return 0;
+        }
+        // for the first frame, initialize OutputsurfaceInfo
+        if (dec_context_->frame_number == 1) {
+            InitOutputFrameInfo(p_frame);
+        }
+        //std::cout<<"Decoding frame: " << dec_context_->frame_number << "<pframe, index>: " << p_frame << " " << av_frame_cnt_ << std::endl;
```
Comment (LakshmiKumar23) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: remove line if unnecessary

File: utils/rocvideodecode/roc_video_dec.h [hunk 24]
```diff
@@ -215,12 +254,12 @@ class RocVideoDecoder {
...
         *   @brief  This function is used to get the current frame size based on pixel format.
         */
-        int GetFrameSize() { assert(disp_width_); return disp_width_ * (disp_height_ + (chroma_height_ * num_chroma_planes_)) * byte_per_pixel_; }
+        virtual int GetFrameSize() { assert(disp_width_); return disp_width_ * (disp_height_ + (chroma_height_ * num_chroma_planes_)) * byte_per_pixel_; }
 
         /**
         *   @brief  This function is used to get the current frame size based on pitch
         */
-        int GetFrameSizePitched() { assert(surface_stride_); return surface_stride_ * (disp_height_ + (chroma_height_ * num_chroma_planes_)); }
+        //int GetFrameSizePitched() { assert(surface_stride_); return surface_stride_ * (disp_height_ + (chroma_height_ * num_chroma_planes_)); }
```
Comment (LakshmiKumar23) [inline utils/rocvideodecode/roc_video_dec.h]: why is this commented out?

File: utils/ffmpegvideodecode/ffmpeg_video_dec.cpp [hunk 25]
```diff
@@ -0,0 +1,866 @@
...
+    if (rocdec_parser_) {
+        rocDecDestroyVideoParser(rocdec_parser_);       // need to do this here since it was already created in the base class
+        // create rocdec videoparser
+        RocdecParserParams parser_params = {};
+        parser_params.codec_type = codec_id_;
+        parser_params.max_num_decode_surfaces = 1; // let the parser to determine the decode buffer pool size
+        parser_params.clock_rate = clk_rate;
+        parser_params.max_display_delay = disp_delay_;
+        parser_params.user_data = this;
+        parser_params.pfn_sequence_callback = FFMpegVideoDecoder::FFMpegHandleVideoSequenceProc;
```
Comment (LakshmiKumar23) [inline utils/ffmpegvideodecode/ffmpeg_video_dec.cpp]: can we call the functions directly since it's member functions of the same class. Do we need the scope?

File: utils/rocvideodecode/roc_video_dec.h (hunk 24 above)
Comment (rrawther) [reply utils/rocvideodecode/roc_video_dec.h]: This function is not used anywhere. removed it

File: samples/videoDecode/videodecode.cpp (hunk 18 above)
Comment (rrawther) [reply samples/videoDecode/videodecode.cpp:217]: Not really needed as it will get reassigned anyway in line#221 onwards

**Review by LakshmiKumar23 — APPROVED** [review]

**Review by AryanSalmanpour — APPROVED** [review]

**Review by spolifroni-amd — APPROVED** [review]
